Restrict custom IO configurations from performing store copy.

maxsumrall opened the pull request:
Custom IO configurations are now supported in single instance mode.
Store copy operations are not supported, and that therefore excludes
using custom IO configurations in HA, online backup, and store migration.
Determining if you violate these restrictions is most easily done by
inspecting config options. This restriction is expected to be short lived and
can be removed once store copy is supported with custom IO configurations.

Commit:
Checks if custom IO configurations conflict with other options.

Custom IO configurations are now supported in single instance mode.
Store copy operations are not supported, and that therefore excludes
using custom IO configurations in HA, online backup, and store migration.
Determining if you violate these restrictions is most easily done by
inspecting config options. This restriction is expected to be short lived and
can be removed once store copy is supported with custom IO configurations.

chrisvest commented on enterprise/backup/src/test/java/org/neo4j/backup/PlatformConstraintBackupTest.java:
```diff
@@ -0,0 +1,58 @@
+        catch ( RuntimeException ex )
+        {
+            // good
+            assertEquals( ("Online Backup not allowed with custom IO integration"), ex.getMessage() );
```
Why the extra parenthesis?

chrisvest commented on enterprise/ha/src/test/java/org/neo4j/graphdb/factory/PlatformConstraintGraphDatabaseFactoryTest.java:
```diff
@@ -0,0 +1,62 @@
+        catch ( RuntimeException ex )
+        {
+            // good
+            assertEquals( ("HA mode not allowed with custom IO integrations"), ex.getMessage() );
```
Parenthesis again

maxsumrall commented:
I'm not sure how those parentheses got there. I removed them.

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -398,6 +398,36 @@ public IndexImplementation apply( String name )
+            {
+                throw new IllegalArgumentException( "Store upgrade not allowed with custom IO integrations" );
+            }
+            String onlineBackup = config.getParams().get( "dbms.backup.enabled" );
```
Can we instead to validation in `EditionModule`s? This will allow to use actual enterprise settings and not hardcode them as strings.

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -398,6 +398,36 @@ public IndexImplementation apply( String name )
+        {
+            if ( config.get( GraphDatabaseSettings.allow_store_upgrade ) )
+            {
+                throw new IllegalArgumentException( "Store upgrade not allowed with custom IO integrations" );
```
Maybe make this and other messages constants and reuse them in tests?

lutovich commented on community/neo4j/src/test/java/upgrade/PlatformConstraintStoreUpgradeTest.java:
```diff
@@ -0,0 +1,58 @@
+        }
+        catch ( RuntimeException ex )
+        {
+            // good
```
Such comments are not really meaningful :)

lutovich commented on community/neo4j/src/test/java/upgrade/PlatformConstraintStoreUpgradeTest.java:
```diff
@@ -0,0 +1,58 @@
+    public TargetDirectory.TestDirectory storeDir = TargetDirectory.testDirForTest( getClass() );
+
+    @Test
+    public void shouldFailToStartIfConfiguredForCAPIDeviceTest()
```
This is not really a `CAPI` specific test. It is more about a custom page swapper. Same applied to other added tests.

Commit:
Fail database startup in incompatible situations.

Custom IO configurations are now supported in single instance mode.
Store copy operations are not supported, and that therefore excludes
using custom IO configurations in HA, online backup, and store migration.
Determining if you violate these restrictions is most easily done by
inspecting config options. This restriction is expected to be short lived and
can be removed once store copy is supported with custom IO configurations.

maxsumrall commented:
@lutovich Thanks, I've tried to address your comments. I think this is a better approach.

Commit:
Consolidate config checking logic and use better exception.
